Copied and modified library macros from /usr/share/autoconf/autoconf/general.m4
with support for using qemu-user-static to execute when cross compiling.

diff --git a/ext/gd/config.m4 b/ext/gd/config.m4
index 7da5b8c..e9dc1dc 100644
--- a/ext/gd/config.m4
+++ b/ext/gd/config.m4
@@ -147,6 +147,40 @@ dnl gdImageCreateFromFoo function even when it does not support the Foo format.
 dnl Those no-op functions display a warning but eventually return normally,
 dnl making a simple link or run test insufficient.
 dnl
+
+AC_DEFUN([_AC_DO_TOKENS_QEMU],
+[{
+  if test "$cross_compiling" = "yes"; then
+    ac_try='QEMU_LD_PREFIX=${XBPS_CROSS_BASE} qemu-${XBPS_TARGET_QEMU_MACHINE}-static $1'
+  else
+    ac_try='$1'
+  fi
+  _AC_DO([$ac_try]); }])
+
+m4_define([_AC_RUN_IFELSE_QEMU_BODY],
+[  AS_LINENO_PUSH([$[]1])
+  AS_IF([_AC_DO_VAR(ac_link) && _AC_DO_TOKENS_QEMU(./conftest$ac_exeext)],
+      [ac_retval=0],
+      [AS_ECHO(["$as_me: program exited with status $ac_status"]) >&AS_MESSAGE_LOG_FD
+       _AC_MSG_LOG_CONFTEST
+       ac_retval=$ac_status])
+  rm -rf conftest.dSYM conftest_ipa8_conftest.oo
+  AS_LINENO_POP
+  AS_SET_STATUS([$ac_retval])
+])
+
+AC_DEFUN([_AC_RUN_IFELSE_QEMU],
+[AC_REQUIRE_SHELL_FN([ac_fn_]_AC_LANG_ABBREV[_try_run_qemu],
+  [AS_FUNCTION_DESCRIBE([ac_fn_]_AC_LANG_ABBREV[_try_run_qemu], [LINENO],
+    [Try to run conftest.$ac_ext, and return whether this succeeded.
+     Assumes that executables *can* be run.])],
+  [$0_BODY])]dnl
+[m4_ifvaln([$1], [AC_LANG_CONFTEST([$1])])]dnl
+[AS_IF([ac_fn_[]_AC_LANG_ABBREV[]_try_run_qemu "$LINENO"], [$2], [$3])
+rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext \
+  conftest.$ac_objext conftest.beam[]m4_ifval([$1], [ conftest.$ac_ext])[]dnl
+])
+
 AC_DEFUN([PHP_GD_CHECK_FORMAT],
 [AS_VAR_PUSHDEF([php_var], [php_cv_lib_gd_gdImageCreateFrom$1])
 old_LIBS="${LIBS}"
@@ -155,7 +189,7 @@ old_CFLAGS="${CFLAGS}"
 CFLAGS="${CFLAGS} ${GDLIB_CFLAGS}"
 AC_LANG_PUSH([C])
 AC_CACHE_CHECK([for working gdImageCreateFrom$1 in libgd], [php_var],
-  [AC_RUN_IFELSE([AC_LANG_SOURCE([
+  [_AC_RUN_IFELSE_QEMU([AC_LANG_SOURCE([
 #include <stdio.h>
 #include <unistd.h>
 #include <gd.h>
@@ -176,7 +210,6 @@ int main(int argc, char** argv) {
   return 0;
 }])],
   [AS_VAR_SET([php_var], [yes])],
-  [AS_VAR_SET([php_var], [no])],
   [AS_VAR_SET([php_var], [no])])])
 AS_VAR_IF([php_var], [yes], [$2])
 AC_LANG_POP([C])
